### PR TITLE
`dfilter` function

### PR DESCRIPTION
**Pull request by MatthewDaggitt**
The current implementation of `filter` in `Data.List` is, to put it mildly, awful:

1. It uses the `Bool` type to filter the contents of the list which encourages the programming style of throwing away type information.
2. It is not even a "native" implementation, instead being a specialisation of `gfilter` being applied to a lengthy lambda. This makes it _very_ difficult to reason about it.

This provides an alternative implementation using a decidable predicate to filter the list which is much cleaner and easier to reason about. Ideally I'd like to ditch the original `filter` but I realise that this would cause a lot of backwards compatibility issues so I think it probably should be reluctantly kept with a warning not to use it...

### CODE REVIEW

**Comment by andreasabel** [comment]
On a very brief look, LGTM.  Thanks!

NB: `gfilter` is actually Haskell's `mapMaybe`.

**Comment by MatthewDaggitt** [comment]
Okay I've updated this PR, as I figured if I'm going to do this I might as well do it right! 

The current implementation of `filter` and `partition` have been renamed `boolFilter` and `boolPartition` and are now marked as deprecated. They have been replaced by new implementations of `filter` and `partition` that use decidable predicates. The function `gfilter` has been renamed to `mapMaybe` to bring it into line with Haskell (thanks @andreasabel!), the old name `gfilter` (and all the corresponding proofs) have been deprecated but kept to maximise backwards compatibility.

Obviously I'm open to any feedback people might have!

I've also noticed that this PR will break the tests for `Issue784` in the main repository, namely:

- LibSucceed/Issue784/Context.agda
- LibSucceed/Issue784/Transformer.agda
- LibSucceed/Issue784/Values.agda

I'm still a little unclear about how the git submodule of the standard library works. Am I correct in thinking that it doesn't automatically pull updates made to the master of this repository? It has to be manually pushed forward right? I guess that means that if I merge this pull request it won't break the tests until the submodule is next updated. 

Is it okay for me to update the submodule immediately when I merge this?

File: CHANGELOG.md [hunk 1]
```diff
@@ -26,8 +26,8 @@ Non-backwards compatible changes
   `boolFilter` and `boolPartition`, and have been replaced by new functions
   `filter` and `partition` which use decidable predicates instead of boolean-valued
   functions.  The former encouraged poor programming style in a dependantly-typed 
-  language such as Agda. Proofs for `boolFilter` and `boolPartition` have also been
-  renamed accordingly.
+  language such as Agda. Proofs for `filter` and `partition` have also been
+  updated and Prenamed accordingly.
```
Comment (andreasabel) [inline CHANGELOG.md]: Is `Prenamed` a typo or some cunning use of language?

**Comment by andreasabel** [comment]
> filter with Bool has not seen many uses in dependently-typed Agda projects

I could be very wrong about this.

File: CHANGELOG.md (hunk 1 above)
Comment (MatthewDaggitt) [reply CHANGELOG.md]: Definitely typo! Will fix.

**Comment by MatthewDaggitt** [comment]
>> filter with Bool has not seen many uses in dependently-typed Agda projects
>
>I could be very wrong about this.

I really hope it's not widely used. It encourages terrible programming style and back when I was first dabbling with Agda, it tied me up in knots trying to convert between predicates and bools and then trying to re-extract the predicates. Not to mention how `boolFilter` is implemented via `mapMaybe`.

It is important enough functionality that it might be present in some code bases. However I think it's worse to leave as is. Without the renaming, a new user might think that the boolean versions are the default versions that you should be using.

I'll leave this PR open for a couple of weeks to see if we get any more feedback. If anyone does know of use-cases, please comment?

**Pull request merged**